File: shopkeeper-angular/src/app/core/services/backup.service.ts
import { Injectable, inject } from '@angular/core';
import { Auth } from '@angular/fire/auth';

import { BusinessStore } from './business-store.service';
import { FirestoreDataService } from './firestore-data.service';
import { ToastService } from './toast.service';
import { BusinessSettings } from '../models';

interface BackupPayload {
  app: 'shopkeeper';
  version: number;
  exportedAt: string;
  settings: BusinessSettings;
  parties: unknown[];
  items: unknown[];
  txns: unknown[];
  adjustments: unknown[];
}

/**
 * Export the signed-in user's full dataset to a JSON file, and restore it back
 * into Firestore. (Data already lives in the cloud and syncs across devices;
 * this is for an offline copy / manual migration.)
 */
@Injectable({ providedIn: 'root' })
export class BackupService {
  private readonly store = inject(BusinessStore);
  private readonly data = inject(FirestoreDataService);
  private readonly auth = inject(Auth);
  private readonly toast = inject(ToastService);

  /** Download all data as a JSON file. */
  exportJson(): void {
    const payload: BackupPayload = {
      app: 'shopkeeper',
      version: 1,
      exportedAt: new Date().toISOString(),
      settings: this.store.settings(),
      parties: this.store.parties(),
      items: this.store.items(),
      txns: this.store.txns(),
      adjustments: this.store.adjustments(),
    };
    const stamp = new Date().toISOString().slice(0, 10);
    const blob = new Blob([JSON.stringify(payload, null, 2)], { type: 'application/json' });
    const a = document.createElement('a');
    a.href = URL.createObjectURL(blob);
    a.download = `shopkeeper-backup-${stamp}.json`;
    document.body.appendChild(a);
    a.click();
    setTimeout(() => { URL.revokeObjectURL(a.href); a.remove(); }, 500);
    this.toast.success('Backup downloaded');
  }

  /** Restore from a backup file, writing every record back to Firestore. */
  async restore(file: File): Promise<void> {
    const uid = this.auth.currentUser?.uid;
    if (!uid) { this.toast.error('Not signed in'); return; }
    let payload: BackupPayload;
    try {
      payload = JSON.parse(await file.text());
    } catch {
      this.toast.error('Invalid backup file');
      return;
    }
    if (payload.app !== 'shopkeeper') { this.toast.error('Not a Shopkeeper backup'); return; }
    try {
      const writes: Promise<void>[] = [];
      for (const p of payload.parties ?? []) writes.push(this.data.upsert(uid, 'parties', p as { id: string }));
      for (const it of payload.items ?? []) writes.push(this.data.upsert(uid, 'items', it as { id: string }));
      for (const t of payload.txns ?? []) writes.push(this.data.upsert(uid, 'txns', t as { id: string }));
      for (const a of payload.adjustments ?? []) writes.push(this.data.upsert(uid, 'adjustments', a as { id: string }));
      if (payload.settings) writes.push(this.data.saveSettings(uid, payload.settings));
      await Promise.all(writes);
      this.toast.success('Backup restored');
    } catch (e) {
      this.toast.error('Restore failed: ' + (e as Error).message);
    }
  }
}

File: shopkeeper-flutter/lib/core/services/backup_service.dart
import 'dart:convert';
import 'dart:io';

import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:file_picker/file_picker.dart';
import 'package:path_provider/path_provider.dart';
import 'package:share_plus/share_plus.dart';

/// Export the user's full dataset to a JSON file (shared via the system sheet)
/// and restore it back into Firestore. Data already lives in the cloud and
/// syncs across devices; this is for an offline copy / manual migration.
class BackupService {
  final FirebaseFirestore _db = FirebaseFirestore.instance;

  Future<void> export(String uid) async {
    final root = _db.collection('userData').doc(uid);
    final payload = <String, dynamic>{
      'app': 'shopkeeper',
      'version': 1,
      'exportedAt': DateTime.now().toIso8601String(),
    };
    for (final col in ['parties', 'items', 'txns', 'adjustments']) {
      final snap = await root.collection(col).get();
      payload[col] = snap.docs.map((d) => d.data()).toList();
    }
    final settings = await root.collection('meta').doc('settings').get();
    payload['settings'] = settings.data() ?? {};

    final dir = await getTemporaryDirectory();
    final stamp = DateTime.now().toIso8601String().substring(0, 10);
    final file = File('${dir.path}/shopkeeper-backup-$stamp.json');
    await file.writeAsString(const JsonEncoder.withIndent('  ').convert(payload));
    await Share.shareXFiles([XFile(file.path)], text: 'Shopkeeper backup');
  }

  /// Returns true on success. Picks a JSON file and writes it back to Firestore.
  Future<bool> restore(String uid) async {
    final result = await FilePicker.platform.pickFiles(type: FileType.custom, allowedExtensions: ['json']);
    if (result == null || result.files.single.path == null) return false;
    final text = await File(result.files.single.path!).readAsString();
    final data = jsonDecode(text) as Map<String, dynamic>;
    if (data['app'] != 'shopkeeper') throw Exception('Not a Shopkeeper backup');

    final root = _db.collection('userData').doc(uid);
    for (final col in ['parties', 'items', 'txns', 'adjustments']) {
      for (final rec in (data[col] as List<dynamic>? ?? [])) {
        final m = Map<String, dynamic>.from(rec as Map);
        await root.collection(col).doc(m['id'].toString()).set(m);
      }
    }
    if (data['settings'] is Map) {
      await root.collection('meta').doc('settings').set(
            Map<String, dynamic>.from(data['settings'] as Map),
            SetOptions(merge: true),
          );
    }
    return true;
  }
}
